doc/developer/design/20240625_source_versioning__table_from_sources.md

```diff
@@ -0,0 +1,577 @@
+   pub storage_metadata: S,
+   /// How to project the ingestion output to this export's collection.
+   pub details: ExportDetails,
+}
+```
+
+Then the subsource-specific values currently stored in `GenericSourceConnection`
+such as the Postgres and MySQL `tables` vecs will be removed, since each
+'source export' would contain its own information.
+
+#### Source Rendering
+
+Currently, when an `IngestionDescription` is used inside `build_ingestion_dataflow`
+(the method that renders the various operators needed to run a source ingestion),
+it determines the `output_index` of each collection to be output by the ingestion
+by calling the `IngestionDescription::source_exports_with_output_indices` method.
+
+This method maps the `external_reference` (the upstream table name) of each `SourceExport`
+to the index of that upstream table in the source's `details` struct (which is currently
+created during purification -- a vector of upstream table descriptions).
+
+Then inside each source implementation, we currently assume that each
+`SourceExport` uniqely corresponds to a single `output_index`, and that each `output_index`
+corresponds to the same index of the upstream table in the source's upstream `details`.
+
+Since the top-level source `details` will no longer contain the `tables` field, the
+output index will be determined by the ordering of the `IngestionDescription::source_exports` `BTreeMap`.
+Each `SourceExport` will output its own stream from the ingestion using
+the `details` it contains. It will be up to each source implementation to map the
+relevant upstream table to the correct `SourceExport`s using the `SourceExportDetails`
+and output to the correct `output_index`.
+
+#### Kafka Sources
+
+Kafka Sources will require more refactoring to move to a model where a single Kafka Source
+can output to more than one `SourceExport`.
+
+Since Kafka Sources only ever refer to a single Kafka Topic, a user will likely only ever
+add a new 'Table' to a Kafka Source to handle a schema change in the messages published
+to that topic.
+
+When a new `SourceExport` is added to a Kafka Source, the new export needs to be hydrated
+from the upstream topic to ensure it has a complete view of the topic contents.
+
+This creates several potential issues regarding the semantics of Kafka Sources:
+
+-   If a new `SourceExport` is added, its resume upper will be the minimum value. If this
+    resume upper is merged with the existing resume upper for the source (which is what
+    happens in Postgres and MySQL sources), it will cause the consumer to begin reading
+    from each partition at the lowest offset available (or the 'start offset' defined
+    on the Kafka Source). This means that we will potentially do a long rehydration
+    to catch up to the previous resume-upper, such that existing `SourceExports` will not
+    see any new data until the rehydration completes.
+-   There is one existing Consumer Group ID used by the Kafka Source, and the source
```

Comment:
@benesch asked this question directly on the commmit - pulling it into the PR discussion:
> I think this one's easy—let's just somehow include the subsource's table's ID in the consumer group ID? Then each subsource will have its progress tracked separately, which seems exactly right.

My response:

ah yes, I think the issue is that `resume_uppers` are provided to our source render implementations as one stream based on the progress collection (cc @petrosagg ), not on a per-collection basis. Additionally it would be a lot less refactoring to continue using a single consumer to read from kafka and then we can just demux to separate outputs when we decode each message.